### PR TITLE
Add constant for 'associated-link-name'

### PR DESCRIPTION
**Pull request by ramya0820**
## Description

Brief description of the changes made in the PR.
- This exposes additional constant required for Service Bus operations.
- Per offline discussions, it is now required to pass in "associated-link-name". Since any link which is idle for 10 mins gets disconnected from the service, passing in the associated link name (say, receiver link on which message was received) will make Service Bus and system not consider the link as idle.

cc @ramya-rao-a @AlexGhiondea

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -225,7 +225,7 @@
     },
     "@types/dotenv": {
       "version": "4.0.3",
-      "resolved": "https://registry.npmjs.org/@types/dotenv/-/dotenv-4.0.3.tgz",
+      "resolved": "http://registry.npmjs.org/@types/dotenv/-/dotenv-4.0.3.tgz",
```
Comment (AlexGhiondea) [inline package-lock.json]: Should the https->http changes be reverted?

/cc @ramya-rao-a

**Comment by ramya0820** [comment]
> The `RequestResponseLink` is a generic concept where we use a AMQP sender and receiver and their session as single entity to send a single request and then get the response.
> 
> Event Hubs and Service Bus use this for $cbs and $management links.
> 
> The need for an `associated-link` is very specific to the request being sent. Not all requests may have an associated-link. Therefore, I don't believe we need to update `SendRequestOptions` for this.

The parameter is that's why made optional. Also ask from Service Bus was to have it on all request-response operations, hence this proposal.

> Why not set the `associated-link` property just like we set any other property when sending the request?
> For example: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/service-bus/src/core/managementClient.ts#L313

Yes, we can do that.
But based on conversation about propagating these changes to all request-response operations, this seems befitting. 
Is there any disadvantage to this approach as such?

File: lib/requestResponseLink.ts [hunk 2]
```diff
@@ -4,8 +4,18 @@
 import * as Constants from "./util/constants";
 import { retry, RetryConfig, RetryOperationType } from "./retry";
 import {
-  Session, Connection, Sender, Receiver, Message as AmqpMessage, EventContext, AmqpError,
-  SenderOptions, ReceiverOptions, ReceiverEvents, ReqResLink, generate_uuid
+  Session,
```
[the remark this answers is not in the document]
Comment (ramya0820) [reply lib/requestResponseLink.ts]: Will do.

**Comment by ramya-rao-a** [comment]
> Is there any disadvantage to this approach as such?

There is no disadvantage as such other than this being not the right place for the fix. The fix for the issue is adding a property to the `application_properties` bag of the request.

- `SendRequestOptions` is the options for how the request is sent and not on the request contents. We are now extending it to contain information on the request contents
- `amqp-common` has by far refrained from editing the contents of the request being made other than adding a messageId if it doesnt exist (probably because it is a required entity). We are now having this library edit the incoming request

> Also ask from Service Bus was to have it on all request-response operations, hence this proposal.

The issue is specific to operations on the $management link. Its still debatable that all request-response operations need this fix. Even if we do arrive at that conclusion, the fix can still be made in the Service Bus SDK such that the fix is restricted to all the $management operations.

**Comment by ramya-rao-a** [comment]
Also, when the service team refers to "all request-response operations" they are referring to the all the operations on the $management link as listed in [this spec ](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-request-response)

**Comment by ramya0820** [comment]
Addressed comments.
@ramya-rao-a @AlexGhiondea 
Can you respond, take a look?

**Comment by ramya-rao-a** [comment]
> Can you respond, take a look?

The points in https://github.com/Azure/amqp-common-js/pull/47#issuecomment-484351956 still hold true. Any updates here?

File: lib/util/constants.ts [hunk 3]
```diff
@@ -67,7 +67,9 @@ export const maxDeadLetterReasonLength = 4096;
 export const maxDurationValue = 922337203685477;
 export const minDurationValue = -922337203685477;
 // https://github.com/Azure/azure-amqp/blob/master/Microsoft.Azure.Amqp/Amqp/AmqpConstants.cs#L47
-export const maxAbsoluteExpiryTime = new Date("9999-12-31T07:59:59.000Z").getTime();
+export const maxAbsoluteExpiryTime = new Date(
```
Comment (ramya0820) [inline lib/util/constants.ts]: This one's a formatting change that got left out on the revert commit

**Comment by ramya0820** [comment]
Thanks for the review @bterlson 

> > Can you respond, take a look?
> 
> The points in [#47 (comment)](https://github.com/Azure/amqp-common-js/pull/47#issuecomment-484351956) still hold true. Any updates here?

About, 

>SendRequestOptions is the options for how the request is sent and not on the request contents. We are now extending it to contain information on the request contents

The statement above seems confusing, as the option added is part of request headers and not content/body.
Do you mean to treat request headers as request content?
The code usage of SendRequestOptions  seems to be packaging request and setting the application properties based on option values, correct?

Overall, it seems like the only concern is appropriateness of fix which is still pending conversations, investigation with team and Service Bus that haven't concluded yet.
For now, since there is no concern about system breaks - would like to move forward with these changes and unblock the bug fix for Azure/azure-sdk-for-js#2288. Does this sound ok? @ramya-rao-a

**Comment by ramya-rao-a** [comment]
> Do you mean to treat request headers as request content?

Yes, when I meant `request content`, I meant everything inside the `request` variable here.

> The code usage of SendRequestOptions seems to be packaging request and setting the application properties based on option values, correct?

2 properties (`delayInSeconds` and `times`) on `SendRequestOptions` are used for the retry logic
3rd property (`timeoutInSeconds`) is used in the timeout that ensures that the request doesnt take more than the configured time

They arent used for packaging the requests or for setting application properties

> Overall, it seems like the only concern is appropriateness of fix which is still pending conversations, investigation with team and Service Bus that haven't concluded yet

The fix (to set the application property) is correct. The concern is where the fix should be made. In `managementClient.ts` file in Service Bus SDK or here.

> For now, since there is no concern about system breaks - would like to move forward with these changes and unblock the bug fix for Azure/azure-sdk-for-js#2288. Does this sound ok? @ramya-rao-a

If unblocking the bug fix is the main priority, then I would prefer a local solution (in Service Bus SDK) rather than one done in an upstream dependency.

**Comment by bterlson** [comment]
FWIW I just chatted with @ramya-rao-a and I now agree with her point about not wanting this change in amqp-common. I thought the associated link was a more common AMQP convention but it's really specific to service bus/event hubs, so I agree amqp-common isn't the right place to make the change. Headers are part of the message contents (similar to HTTP where headers are considered part of a request/response).

**Comment by ramya0820** [comment]
Got it, thanks for clarifying - please take a look again @bterlson @ramya-rao-a 
Based on notes from @AlexGhiondea updated description to not be a blocker for issue# 2288 in azure-sdk-for-js repo, and tracking this as part of https://github.com/Azure/azure-sdk-for-js/issues/2361 instead

**Comment by ramya-rao-a** [comment]
Thanks @ramya0820

**Comment by ramya0820** [comment]
Can anyone merge this PR?

@bterlson @ramya-rao-a @AlexGhiondea

**Pull request merged**